pandas/core/nanops.py

```diff
@@ -254,9 +254,16 @@ def nansum(values, axis=None, skipna=True):
 @bottleneck_switch()
 def nanmean(values, axis=None, skipna=True):
     values, mask, dtype, dtype_max = _get_values(values, skipna, 0)
-    the_sum = _ensure_numeric(values.sum(axis, dtype=dtype_max))
     count = _get_counts(mask, axis)
 
+    dtype_sum = dtype_max
+    if is_integer_dtype(dtype):
+        dtype_sum = np.float64
+    elif is_float_dtype(dtype):
+        dtype_sum = dtype
+        count = dtype.type(count)
```

Comment:
oh that's because for `float` dtypes `count` can be higher precision and therefore changing the dtype of the returned result. E.g. for `float32` input count is still `float64` and therefore the result would be cast to `float64` when we do `sum / count`, and that's not what we want